src/frontends/tensorflow/src/helper_transforms/block_lstm_replacer.cpp

```diff
@@ -18,20 +18,72 @@
 
 using namespace std;
 using namespace ov::pass;
+using namespace ov::pass::pattern;
 using namespace ov::opset9;
 using namespace ov::frontend::tensorflow;
 
-pass::BlockLSTMToLSTMSequenceOneOutput::BlockLSTMToLSTMSequenceOneOutput() {
-    auto block_lstm = pattern::wrap_type<BlockLSTM>();
+namespace {
+std::function<bool(ov::Output<ov::Node>)> can_have_outputs(const std::vector<size_t>& allowed_output_indices) {
+    return [=](ov::Output<ov::Node> output) -> bool {
+        auto block_lstm_node = std::dynamic_pointer_cast<BlockLSTM>(output.get_node_shared_ptr());
+        if (!block_lstm_node) {
+            return false;
+        }
+        auto output_size = block_lstm_node->get_output_size();
+        for (size_t output_ind = 0; output_ind < output_size; ++output_ind) {
+            if (std::find(allowed_output_indices.begin(), allowed_output_indices.end(), output_ind) !=
+                allowed_output_indices.end()) {
+                continue;
+            }
+            if (block_lstm_node->output(output_ind).get_target_inputs().size() > 0) {
+                return false;
+            }
+        }
+        return true;
+    };
+}
+}  // namespace
 
-    matcher_pass_callback callback = [=](pattern::Matcher& m) {
-        NodeRegistry rg;
+pass::BlockLSTMReplacer::BlockLSTMReplacer() {
+    // Pattern 1: BlockLSTM with last state cell output (BlockLSTM -> Concat -> GatherND)
+    // used in DeepSpeech model
+    auto block_lstm_1 = pattern::wrap_type<BlockLSTM>(can_have_outputs({1, 6}));
+    auto states_cell_1 = pattern::wrap_type<Concat>({pattern::any_input(), block_lstm_1});
+    auto pattern1 = pattern::wrap_type<GatherND>({states_cell_1, pattern::any_input()});
+
+    // Pattern 2: BlockLSTM with just one output, concatenated hidden states (BlockLSTM)
+    auto pattern2 = pattern::wrap_type<BlockLSTM>(can_have_outputs({6}));
 
-        auto block_lstm_node = std::dynamic_pointer_cast<BlockLSTM>(m.get_match_root());
+    auto root = std::make_shared<pattern::op::Or>(OutputVector{pattern1, pattern2});
+
+    matcher_pass_callback callback = [=](pattern::Matcher& m) {
+        auto pattern_map = m.get_pattern_value_map();
+        auto is_pattern1 = (pattern_map.find(pattern1) != std::end(pattern_map));
+        auto is_pattern2 = (pattern_map.find(pattern2) != std::end(pattern_map));
+
+        // find for each pattern BlockLSTM node for which we adjust inputs
+        // and check its attributes before the transformation
+        std::shared_ptr<BlockLSTM> block_lstm_node = nullptr;
+        std::shared_ptr<Node> last_state_c_node = nullptr;
+        ov::NodeVector rt_info_from;
+        if (is_pattern1) {
+            block_lstm_node = std::dynamic_pointer_cast<BlockLSTM>(pattern_map.at(block_lstm_1).get_node_shared_ptr());
```

Comment:
it looks like it's not necessary to cast (std::dynamic_pointer_cast) this object
you don't use any specific methods of this class

Reply:
I need it because below I use:

```
auto hidden_size = block_lstm_node->get_hidden_size();
```